### DISCUSSION TITLE
no such overload: !_

### DISCUSSION THREAD

**Discussion by davenewza** [google/cel-go#1058] [Q&A]
With v0.21.0, I'm using `NewCustomEnv` with custom declarations for just the basic boolean operators `||`, `&&` and `!`.  However, evaluation fails when I use `!` in an expression (e.g. `!value`):

> no such overload: !_

`||` and `&&` work just fine. 

```go
env, err := cel.NewCustomEnv(
	cel.Declarations(
		decls.NewFunction(operators.LogicalAnd, decls.NewOverload(
			overloads.LogicalAnd,
			[]*expr.Type{decls.Bool, decls.Bool},
			decls.Bool,
		)),
		decls.NewFunction(operators.LogicalOr, decls.NewOverload(
			overloads.LogicalOr,
			[]*expr.Type{decls.Bool, decls.Bool},
			decls.Bool,
		)),
		decls.NewFunction(operators.LogicalNot, decls.NewOverload(
			overloads.LogicalNot,
			[]*expr.Type{decls.Bool},
			decls.Bool,
		))),
	cel.Variable("value", cel.BoolType),
)
if err != nil {
	log.Fatalf("program setup err: %s", err)
}

ast, issues := env.Compile("!value")
if issues != nil && issues.Err() != nil {
	log.Fatalf("type-check error: %s", issues.Err())
}

prg, err := env.Program(ast)
if err != nil {
	log.Fatalf("program construction error: %s", err)
}

out, _, err := prg.Eval(map[string]interface{}{
	"value": true,
})
if err != nil {
	log.Fatalf("eval error: %s", err) // <--- fails here
}

fmt.Println(out) // <- expecting 'false'
```

Can someone please point out what I am doing wrong.

**Comment by TristonianJones** [comment]
Hi @davenewza,

You need to provide the implementation of your version of `!`. Since `||` and `&&` are so integral to CEL and have very special semantics we don't allow overloads of these functions (which means they magically exist without further configuration). If you want to provide your own logical not, you'd need to look at how we do this in the `common/stdlib` package: 

https://github.com/google/cel-go/blob/f9db1d6dab3ebe7e2d885dfb00db976254df1faf/common/stdlib/standard.go#L69-L77

**Reply by davenewza** [reply]
Thanks @TristonianJones. I discovered so upon further experimenting.  I had assumed `!` to have "native" support.
